Incorporate exportloopref into tests

The general category of bug that caused the CAA Rechecking bug (https://bugzilla.mozilla.org/show_bug.cgi?id=1619047) is "[using a reference to a loop iterator variable](https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable)." This tool aims to detect such cases: https://github.com/kyoh86/exportloopref. It wouldn't have caught our specific bug, which falls under the "known false negatives" for the tool: "Case 1: pass the pointer to function to export." But I still think it's a potentially valuable tool to catch other mistakes in the general category.

Initial investigation:

At the moment, exportloopref has no findings for Boulder, and (as you note) it wouldn't have caught the CAA Rechecking bug:
```
± git log --oneline HEAD~..HEAD
8923fa29 (HEAD, origin/master, origin/main, origin/HEAD) Remove infinite memlock rlimit from docker config (#4967)
± ~/go/bin/exportloopref ./... 
± echo $?                      
0
± git log --oneline aa01f4d0~..aa01f4d0
aa01f4d0 (tag: release-2020-02-28) Pass authzModel by value, not reference (#4690)
± git reset --hard aa01f4d0~
HEAD is now at 542cb6d2 Make orders with deact authz invalid (#4687)
± ~/go/bin/exportloopref ./...         
± echo $?
0
± git stash pop
<snip>
± git diff
diff --git a/ca/ca.go b/ca/ca.go
index 566893eb..f775cd58 100644
--- a/ca/ca.go
+++ b/ca/ca.go
@@ -234,6 +234,11 @@ func NewCertificateAuthorityImpl(
                return nil, err
        }
 
+       var ints []*int
+       for _, i := range []int{1, 2, 3, 4} {
+               ints = append(ints, &i)
+       }
+
        if config.LifespanOCSP.Duration == 0 {
                return nil, errors.New("Config must specify an OCSP lifespan period.")
        }
± ~/go/bin/exportloopref ./...
~/Code/boulder/ca/ca.go:239:24: exporting a pointer for the loop variable i
± echo $?
3
```

But getting it running isn't that hard. It doesn't play _super_ nicely with the module system (if we add it to go.mod it still won't be fetched, vendored, or built because nothing actually imports it). But that should be relatively straightforward to work around.

--- a/.golangci.yml
+++ b/.golangci.yml
@@ -15,24 +15,20 @@ issues:
   exclude-rules:
     - linters:
       - gosec
-        # TODO: Identify, fix, and remove violations of each of these rules
-        # (sorted in order of decreasing frequency in our codebase).
-        # G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
-        # G304 (CWE-22) : Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
-        # G404 (CWE-338): Use of weak random number generator (math/rand instead of crypto/rand) (Confidence: MEDIUM, Severity: HIGH)
-        # G401 (CWE-326): Use of weak cryptographic primitive (Confidence: HIGH, Severity: MEDIUM)
-        # G306 (CWE-276): Expect WriteFile permissions to be 0600 or less (Confidence: HIGH, Severity: MEDIUM)
-        # G505 (CWE-327): Blocklisted import crypto/sha1: weak cryptographic primitive (Confidence: HIGH, Severity: MEDIUM)
-        # G601 (CWE-118): Implicit memory aliasing in for loop. (Confidence: MEDIUM, Severity: MEDIUM)
-        # G402 (CWE-295): TLS MinVersion too low. (Confidence: HIGH, Severity: HIGH)
-        # G402 (CWE-295): TLS InsecureSkipVerify set true. (Confidence: HIGH, Severity: HIGH)
-        # G204 (CWE-78) : Subprocess launched with variable (Confidence: HIGH, Severity: MEDIUM)
-        # G402 (CWE-295): TLS MaxVersion too low. (Confidence: HIGH, Severity: HIGH)
-        # G501 (CWE-327): Blocklisted import crypto/md5: weak cryptographic primitive (Confidence: HIGH, Severity: MEDIUM)
-        # G403 (CWE-310): RSA keys should be at least 2048 bits (Confidence: HIGH, Severity: MEDIUM)
-        # G307 (CWE-703): Deferring unsafe method "Close" on type "*os.File" (Confidence: HIGH, Severity: MEDIUM)
-        # G107 (CWE-88) : Potential HTTP request made with variable url (Confidence: MEDIUM, Severity: MEDIUM)
-      text: "(G104|G304|G404|G401|G306|G505|G601|G402|G204|G501|F403|G307|G107)"
+      # TODO: Identify, fix, and remove violations of each of these rules
+      # G101: Potential hardcoded credentials
+      # G102: Binds to all network interfaces
+      # G107: Potential HTTP request made with variable url
+      # G201: SQL string formatting
+      # G202: SQL string concatenation
+      # G306: Expect WriteFile permissions to be 0600 or less
+      # G401: Use of weak cryptographic primitive
+      # G402: TLS InsecureSkipVerify set true.
+      # G403: RSA keys should be at least 2048 bits
+      # G501: Blacklisted import `crypto/md5`: weak cryptographic primitive
+      # G505: Blacklisted import `crypto/sha1`: weak cryptographic primitive
+      # G601: Implicit memory aliasing in for loop.
+      text: "G(101|102|107|201|202|306|401|402|403|501|505|601)"
     - linters:
       - staticcheck
       text: "(SA1019|ST1005|ST1013|SA6003|SA5011|S1029|SA2002):"

--- a/.golangci.yml
+++ b/.golangci.yml
@@ -1,10 +1,11 @@
 linters:
   disable-all: true
   enable:
-    - govet
-    - gofmt
-    - ineffassign
     - errcheck
+    - gofmt
+    - gosec
+    - govet
+    - ineffassign
     - misspell
     - staticcheck
 linters-settings:
@@ -12,6 +13,26 @@ linters-settings:
     ignore: fmt:[FS]?[Pp]rint*,io:Write,os:Remove,net/http:Write,github.com/miekg/dns:WriteMsg,net:Write,encoding/binary:Write
 issues:
   exclude-rules:
+    - linters:
+      - gosec
+        # TODO: Identify, fix, and remove violations of each of these rules
+        # (sorted in order of decreasing frequency in our codebase).
+        # G104 (CWE-703): Errors unhandled. (Confidence: HIGH, Severity: LOW)
+        # G304 (CWE-22) : Potential file inclusion via variable (Confidence: HIGH, Severity: MEDIUM)
+        # G404 (CWE-338): Use of weak random number generator (math/rand instead of crypto/rand) (Confidence: MEDIUM, Severity: HIGH)
+        # G401 (CWE-326): Use of weak cryptographic primitive (Confidence: HIGH, Severity: MEDIUM)
+        # G306 (CWE-276): Expect WriteFile permissions to be 0600 or less (Confidence: HIGH, Severity: MEDIUM)
+        # G505 (CWE-327): Blocklisted import crypto/sha1: weak cryptographic primitive (Confidence: HIGH, Severity: MEDIUM)
+        # G601 (CWE-118): Implicit memory aliasing in for loop. (Confidence: MEDIUM, Severity: MEDIUM)
+        # G402 (CWE-295): TLS MinVersion too low. (Confidence: HIGH, Severity: HIGH)
+        # G402 (CWE-295): TLS InsecureSkipVerify set true. (Confidence: HIGH, Severity: HIGH)
+        # G204 (CWE-78) : Subprocess launched with variable (Confidence: HIGH, Severity: MEDIUM)
+        # G402 (CWE-295): TLS MaxVersion too low. (Confidence: HIGH, Severity: HIGH)
+        # G501 (CWE-327): Blocklisted import crypto/md5: weak cryptographic primitive (Confidence: HIGH, Severity: MEDIUM)
+        # G403 (CWE-310): RSA keys should be at least 2048 bits (Confidence: HIGH, Severity: MEDIUM)
+        # G307 (CWE-703): Deferring unsafe method "Close" on type "*os.File" (Confidence: HIGH, Severity: MEDIUM)
+        # G107 (CWE-88) : Potential HTTP request made with variable url (Confidence: MEDIUM, Severity: MEDIUM)
+      text: "(G104|G304|G404|G401|G306|G505|G601|G402|G204|G501|F403|G307|G107)"
     - linters:
       - staticcheck
       text: "(SA1019|ST1005|ST1013|SA6003|SA5011|S1029|SA2002):"

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -2,7 +2,7 @@ version: '3'
 services:
     boulder:
         # To minimize fetching this should be the same version used below
-        image: letsencrypt/boulder-tools-go${TRAVIS_GO_VERSION:-1.14.5}:2020-07-14
+        image: letsencrypt/boulder-tools-go${TRAVIS_GO_VERSION:-1.14.5}:2020-07-28
         environment:
             - FAKE_DNS=10.77.77.77
             - BOULDER_CONFIG_DIR=test/config
@@ -76,7 +76,7 @@ services:
         logging:
             driver: none
     netaccess:
-        image: letsencrypt/boulder-tools-go${TRAVIS_GO_VERSION:-1.14.5}:2020-07-14
+        image: letsencrypt/boulder-tools-go${TRAVIS_GO_VERSION:-1.14.5}:2020-07-28
         environment:
             GO111MODULE: "on"
             GOFLAGS: "-mod=vendor"

--- a/test/boulder-tools/build.sh
+++ b/test/boulder-tools/build.sh
@@ -30,7 +30,7 @@ unzip /tmp/protoc.zip -d /usr/local/protoc
 export GOBIN=/usr/local/bin GOCACHE=/tmp/gocache
 
 # Install golangci-lint
-curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $GOBIN v1.24.0
+curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $GOBIN v1.29.0
 
 # Install protobuf and testing/dev tools.
 # Note: The version of golang/protobuf is partially tied to the version of grpc

